Make it possible to parse large messages with ML-DSA

simo5 opened the pull request:
This PR split mldsa sign/verify functions so that messages can be fed via an update() interface...

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

simo5 commented:
Still working on this, but would like early feedback if anyone has strong opinions

slontis commented:
I think Pauli was thinking of looking at this also..

Potentially you could avoid most of the storage you are doing for the ctx..
You only need tr & M' to calculate mu.
There would be nothing stopping you making a skDecode variant that just extracted tr (probably just check the size is correct).

The verify case is even easier i.e. only  tr = H(pk,64) is required to calculate mu.

paulidale commented:
I'll try to get a look at this today.  I was thinking of doing the same thing :grin:

simo5 commented:
> I think Pauli was thinking of looking at this also..
> 
> Potentially you could avoid most of the storage you are doing for the ctx.. You only need tr & M' to calculate mu.
> 
> So the init could get tr - Hash it, then hash the encoded ctx. The update then just hashes the message Final() then does everything other than calculate mu.
> 
> There would be nothing stopping you making a skDecode variant that just extracted tr (probably just check the size is correct).
> 
> The verify case is even easier i.e. only tr = H(pk,64) is required to calculate mu.

Yeah I think this is definitely not a bad idea, I will look into deferring as many operations as possible.

simo5 commented:
So the more I look at this split now, the more I am thinking that perhaps the low level functions should be split among a different line.

Provide a single set of functions to compute mu that has init/update/final, and then the actual sign/verify functions as one shot.

I will proceed in that direction with the next update to this PR

simo5 commented:
Added tests, and this, I think, concludes the necessary changes.

slontis commented on crypto/ml_dsa/ml_dsa_sign.c:
```diff
@@ -43,12 +43,110 @@ static void signature_init(ML_DSA_SIG *sig,
 /*
- * FIPS 204, Algorithm 7, ML-DSA.Sign_internal()
- * @returns 1 on success and 0 on failure.
+ * @brief: Auxiliar functions to compute ML-DSA's MU.
```
Auxiliary ?

simo5 replied:
Please suggest any wording that works for you.

t8m replied:
There is just a typo auxiliar -> auxiliary

simo5 replied:
aahh, sorry I am slightly dislexic, sometimes I do not see those typos at all :)

slontis commented on crypto/ml_dsa/ml_dsa_sign.c:
```diff
@@ -43,12 +43,110 @@ static void signature_init(ML_DSA_SIG *sig,
+    EVP_MD_CTX *md_ctx;
+    uint8_t itb[2];
+
+    if (!key)
```
key == NULL

slontis commented on doc/man3/EVP_PKEY_verify.pod:
```diff
@@ -50,31 +50,31 @@ I<ctx> for verifying an unlimited size message using the algorithm given by
-It I<must> be used together with EVP_PKEY_verify_update() and
-EVP_PKEY_verify_final().
+I<sig> to be verified against by EVP_PKEY_verify_message_final().
+It I<must> be used together with EVP_PKEY_verify_message_update() and
```
Not sure if these ones should be changed..

simo5 replied:
There is no function called EVP_PKEY_verify_update() or EVP_PKEY_verify_final(), why wouldn't you want to fix this issue ?

slontis commented on providers/implementations/signature/ml_dsa_sig.c:
```diff
@@ -54,13 +58,20 @@ typedef struct {
+    int operation;
+    EVP_MD_CTX *md_ctx; /* Ctx for msg_init/update/final interface */
+    unsigned char *sig; /* Signature, for verification */
+    size_t siglen;
```
be consistent with rest of file sig_len.

simo5 replied:
most functions take "siglen", and this is the same name that exist in PROV_RSA_CTX and PROV_ECDSA_CTX so I tried to stay consistent with the other implementations.

slontis commented on providers/implementations/signature/ml_dsa_sig.c:
```diff
@@ -92,7 +104,32 @@ static void *ml_dsa_dupctx(void *vctx)
+        }
+    }
+
+    if (srcctx->md_ctx) {
```
!= NULL

slontis commented on providers/implementations/signature/ml_dsa_sig.c:
```diff
@@ -281,12 +404,26 @@ static int ml_dsa_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+    PROV_ML_DSA_CTX *pctx = (PROV_ML_DSA_CTX *)vctx;
+
     static const OSSL_PARAM settable_ctx_params[] = {
         OSSL_PARAM_octet_string(OSSL_SIGNATURE_PARAM_CONTEXT_STRING, NULL, 0),
```
Maybe macro the shared ones as a #define and include them in both..

simo5 replied:
ok (boy do I hate all these macros, makes reading the code sooo hard)

slontis reviewed (commented):
Looks pretty clean after the rework.. good job.

slontis reviewed (commented):
Should this have Example code in EVP_SIGNATURE-ML-DSA.pod?

simo5 commented:
Rebased on master, squashed together the main split commit and the original commit from #27322 (which has now been closed), merged all the fixups in the respective commits.

t8m reviewed (changes requested):
Great work! Some nits found.

simo5 commented on crypto/ml_dsa/ml_dsa_sign.c:
```diff
@@ -43,12 +43,110 @@ static void signature_init(ML_DSA_SIG *sig,
-                                size_t encoded_msg_len,
+int ossl_ml_dsa_mu_finalize(EVP_MD_CTX *md_ctx, uint8_t *mu, size_t mu_len)
+{
+    if (mu_len != ML_DSA_MU_BYTES)
```
The API is now internally public, it may be a real failure if used more extensively in future, for example if some of this is later shared with SLH-DSA (ML-DSA and SLH-DSA do not do exactly the same things, but ...)

t8m replied:
ossl_assert just makes it an assert in debug builds but normal condition in regular builds, IMO it can still be used.

simo5 commented:
> Great work! Some nits found.

I think I addressed all of those, PTAL

t8m commented on crypto/ml_dsa/ml_dsa_sign.c:
```diff
@@ -126,8 +129,10 @@ int ossl_ml_dsa_mu_update(EVP_MD_CTX *md_ctx, const uint8_t *msg, size_t msg_len
 int ossl_ml_dsa_mu_finalize(EVP_MD_CTX *md_ctx, uint8_t *mu, size_t mu_len)
 {
-    if (mu_len != ML_DSA_MU_BYTES)
+    if (!ossl_assert(mu_len != ML_DSA_MU_BYTES)) {
```
You need to invert this condition as there must be negation before the ossl_assert() call.

t8m commented on providers/implementations/signature/ml_dsa_sig.c:
```diff
@@ -174,6 +212,72 @@ static int ml_dsa_digest_signverify_init(void *vctx, const char *mdname,
+                return 0;
+        }
+
+        if (!ossl_ml_dsa_mu_finalize(ctx->md_ctx, mu, ML_DSA_MU_BYTES))
```
Could you please replace the constant with sizeof also here?

t8m commented on providers/implementations/signature/ml_dsa_sig.c:
```diff
@@ -213,10 +317,28 @@ static int ml_dsa_digest_sign(void *vctx, uint8_t *sig, size_t *siglen, size_t s
+    if (ctx->md_ctx == NULL)
+        return 0;
+
+    if (!ossl_ml_dsa_mu_finalize(ctx->md_ctx, mu, ML_DSA_MU_BYTES))
```
And here?

simo5 commented:
I am not sure how I broke the FIPS compatibility tests with this patch... is it a false positive somehow?

simo5 commented:
Ah I guess the 3.5.0 FIPS provider does not support the message based signature updates, how do I mark that test to be incompatible with the 3.5.0 provider ?

slontis commented:
for ml_dsa_test you can use fips_provider_version_ge() of fips_provider_version_match..

For the evp_test related ones you have to add metadata to the txt file entries (either first or after the Availablein line).
e.g. FIPSversion = >=3.6.0

t8m commented:
> I am not sure how I broke the FIPS compatibility tests with this patch... is it a false positive somehow?

Not a false positive unfortunately - the testcase will be run with a FIPS provider from the 3.5 branch and that won't support the new functionality. So you need to `TEST_skip()` that testcase if `fips_provider_version_lt(lib_ctx, 3, 6, 0)`

t8m commented:
BTW, only the test_mldsa failure vs 3.5 branch is relevant, the others are a separate regression.

t8m commented:
> For the evp_test related ones you have to add metadata to the txt file entries (either first or after the Availablein line).
> e.g. FIPSversion = >=3.6.0

This is an unrelated regression, I'm looking into it.

vdukhovni reviewed (approved):
This is great work, but I was also hoping we'd be able to support IUF for pure ML-DSA even with `EVP_DigestSignInit()`, and friends, provided the digest is either NULL or happens to be SHAKE256, with the calls to the new provider methods handled internally.

That would make it possible to perform ML-DSA signing of large messages without switching to a new API...

Perhaps that could be done in a followup PR?

simo5 commented:
> This is great work, but I was also hoping we'd be able to support IUF for pure ML-DSA even with `EVP_DigestSignInit()`, and friends, provided the digest is either NULL or happens to be SHAKE256, with the calls to the new provider methods handled internally.
> 
> That would make it possible to perform ML-DSA signing of large messages without switching to a new API...
> 
> Perhaps that could be done in a followup PR?

I thought about this too, and want this too, but definitely in a followup PR.
I would also like to add direct support for HashML-DSA, rather than having people have to use the external MU thing to cobble it up together themselves ...

vdukhovni commented on crypto/ml_dsa/ml_dsa_sign.c:
```diff
@@ -258,15 +371,7 @@ static int ml_dsa_verify_internal(const ML_DSA_KEY *pub, int msg_is_mu,
-            goto err;
-    }
+
     /* Compute verifiers challenge c_ntt = NTT(SampleInBall(c_tilde) */
```
```suggestion
    /* Compute verifiers challenge c_ntt = NTT(SampleInBall(c_tilde)) */
```

simo5 replied:
yes, this one needs closing :)

vdukhovni replied:
FWIW, though I am not forcing the issue, I don't think that leaving the open paren is a particularly effective way to outline the logic.  Instead just just list what's going to be hashed in the comment, or refer to the appropriate FIPS 203 section, and then just let the code clearly match the comment.

simo5 replied:
The general M' logic is already explained in the function description, if you want me to remove the comments thru the function I can do that, I have no attachments to code.

vdukhovni replied:
Do what you think is best, but if the comments are redundant, feel free to drop them, but keep them if they're useful...

vdukhovni reviewed (approved):
LGTM, with a suggestion to balance parens in comments.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to the master branch. Thank you for your contribution.
